### PR TITLE
414 Added missing endpoints to resume and portfolio api 

### PR DESCRIPTION
**Pull request by ThunderIW**
## 📝 Description
- Added missing endpoints for  Portfolio Generator API and Resume Generator API endpoints 
- Added missing test cases for the Portfolio Generator API and Resume Generator API endpoints. The existing test suites did not 
  

**Closes:** #441  

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

   ## 📦 Why This PR Is Large

  This PR appears large due to several factors beyond the core feature work:

  1. **New endpoints added:** `Portfolio_Generator_API.py` and `Resume_Generator_API.py` were extended with missing CRUD
  endpoints for education, experience, project removal, default render, and export.
  2. **Test coverage expanded:** Test coverage expanded: 16 new test methods were added across 9 new test classes covering all previously untested endpoints and error paths.

  **Actual new code added in this PR:** 577 lines across 4 files.

  | File | Change | Lines Added |
  |------|--------|-------------|
  | `src/API/Portfolio_Generator_API.py` | Modified — added default render, export endpoints | 22 |
  | `src/API/Resume_Generator_API.py` | Modified — added education/experience CRUD, remove project, default render, export endpoints | 181 |
  | `test/test_portfolio_generator_API.py` | Updated — added 5 new test methods across 3 classes | 105 |
  | `test/test_resume_generator_API.py` | Updated — added 11 new test methods across 6 classes | 267 |

---

## 🧪 Testing

> All 53 tests pass via

Testing Steps

local pytest
```
python -m pytest test/test_portfolio_generator_API.py test/test_resume_generator_API.py -v
```

Docker Pytest
```bash
docker-compose run test python -m pytest test/test_portfolio_generator_API.py test/test_resume_generator_API.py
```

### Portfolio API Tests (`test_portfolio_generator_API.py`)

| Test Class | Test Method | What It Covers |
|---|---|---|
| `TestRemoveProject` | `test_success_and_error_cases` | DELETE project success, project not found (404), no projects (404), portfolio not found (404) |
| `TestAddProjectExtended` | `test_payload_overrides_and_check_result_failure` | Add project with payload overrides succeeds; `_check_result` rejects non-success string (400) |
| `TestAddProjectExtended` | `test_default_render_and_showcase_save_error` | POST `/render` (no format) defaults to PDF with correct headers |
| `TestAddProjectExtended` | `test_showcase_role_save_error_returns_500` | `save_project_role_override` exception returns 500 |
| `TestExportPortfolioErrors` | `test_export_not_found_cases` | Default and custom export return 404 for missing portfolios |

### Resume API Tests (`test_resume_generator_API.py`)

| Test Class | Test Method | What It Covers |
|---|---|---|
| `TestGenerateAndEditEdgeCases` | `test_generate_invalid_theme_and_default_skips_update` | Invalid theme returns 400; default theme `sb2nov` skips `update_theme` |
| `TestGenerateAndEditEdgeCases` | `test_edit_invalid_contact_field_and_theme_and_check_result` | Invalid contact field (400), invalid theme in edit (400), `_check_result` failure (400) |
| `TestGenerateAndEditEdgeCases` | `test_default_render_returns_pdf` | POST `/render` (no format) defaults to PDF |
| `TestGenerateAndEditEdgeCases` | `test_delete_os_error_returns_500` | OSError during resume deletion returns 500 |
| `TestRemoveProject` | `test_success_and_error_cases` | DELETE project success, project not found (404), no projects (404), resume not found (404) |
| `TestEducationEndpoints` | `test_add_and_remove_education` | Add success, duplicate (409), bad result (400), remove success, remove not found (404), empty education list (404) |
| `TestEducationEndpoints` | `test_resume_not_found` | Add and remove education both return 404 for missing resume |
| `TestExperienceEndpoints` | `test_add_and_remove_experience` | Add success, add failure (400), remove success, remove not found (404), empty experience list (404) |
| `TestExperienceEndpoints` | `test_resume_not_found` | Add and remove experience both return 404 for missing resume |
| `TestAddProjectExtended` | `test_payload_overrides_and_check_result_failure` | Payload overrides DB values; `_check_result` rejects non-success string (400) |
| `TestExportResumeErrors` | `test_export_not_found_cases` | Default and custom export return 404 for missing resumes |

### Local Pytest result
image

### Docker pytest result
image

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [ ] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

</details>

### CODE REVIEW

File: src/API/Resume_Generator_API.py [hunk 1]
```diff
@@ -489,6 +536,134 @@ def add_project(id: str, project_name: str, payload: Optional[ProjectRequest] =
...
+
+    Returns:
+        dict: {"status": str} confirming the education was removed.
+
+    Raises:
+        HTTPException: 404 if the resume or institution does not exist.
+    """
+    doc = _load_resume(id)
+    result = doc.remove_education(institution_name)
+    if "not found" in result.lower():
```
Comment (Puneet-Maan) [inline src/API/Resume_Generator_API.py]: if doc.remove_education() returns something like "No education to delete", this still returns 200 right now. Can we treat that as 404 too, same as not found?

File: src/API/Resume_Generator_API.py [hunk 2]
```diff
@@ -489,6 +536,134 @@ def add_project(id: str, project_name: str, payload: Optional[ProjectRequest] =
...
+
+    Returns:
+        dict: {"status": str} confirming the experience was removed.
+
+    Raises:
+        HTTPException: 404 if the resume or company does not exist.
+    """
+    doc = _load_resume(id)
+    result = doc.remove_experience(company_name)
+    if "not found" in result.lower():
```
Comment (Puneet-Maan) [inline src/API/Resume_Generator_API.py]: Same thing here for empty experience list (No experience to delete), currently this comes back as 200. Can we map that to 404 as well?

File: test/test_resume_generator_API.py [hunk 3]
```diff
@@ -234,6 +234,261 @@ def test_render_output_parent_directory_cleanup(self, _mock_shutil):
...
+        self._set_not_found()
+        resp = self.client.delete("/resume/fake_id/project/MyProject")
+        self.assertEqual(resp.status_code, 404)
+
+
+class TestEducationEndpoints(_BaseResumeTest):
+    """Tests for POST /resume/{id}/add/education and DELETE /resume/{id}/education/{name}."""
+
+    def test_add_and_remove_education(self):
+        """Covers add success, duplicate (409), bad result (400), remove success, and remove not found (404)."""
```
Comment (Puneet-Maan) [inline test/test_resume_generator_API.py]: Can we add one test for the "No education to delete" path?

**Review by Puneet-Maan — CHANGES_REQUESTED** [review]
Gooood work in adding these missing pieces like endpoints, tests, etc.
There's two edge cases in resume delete endpoints where empty lists may return 200 instead of 404 (education, experience).
If you could  patch those and then add 2 tests for it, please.

File: test/test_resume_generator_API.py (hunk 3 above)
Comment (ThunderIW) [reply test/test_resume_generator_API.py]: > Can we add one test for the "No education to delete" path?

For sure
